### PR TITLE
compiler: Patch CSE in presence of conditionals

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

**Review by georgebisbas — APPROVED** [review]

File: devito/passes/clusters/cse.py [hunk 1]
```diff
@@ -98,24 +98,26 @@ def _cse(maybe_exprs, make, min_cost=1, mode='default'):
...
-                pe, changed = _uxreplace(pe, {t.rhs: t.lhs})
-                if changed and t not in updated:
-                    updated.append(t)
+            for k, v in chosen:
+                pe, changed = _uxreplace(pe, {k: v})
+                if changed and v not in scheduled:
+                    updated.append(pe.func(v, k, operation=None))
+                    scheduled.append(v)
             updated.append(pe)
         processed = updated
```
Comment (EdCaunt) [inline devito/passes/clusters/cse.py:116]: Why not just use `updated` henceforth, rather than using `processed`?

File: tests/test_dse.py [hunk 2]
```diff
@@ -168,6 +168,29 @@ def test_cse_temp_order():
     assert type(args[2]) is CTemp
 
 
+def test_cse_w_conditionals():
+    grid = Grid(shape=(10, 10, 10))
+    x, _, _ = grid.dimensions
```
Comment (EdCaunt) [inline tests/test_dse.py:173]: Nitpick: `x = grid.dimensions[0]` would be tidier imo

Comment (FabioLuporini) [reply tests/test_dse.py:173]: tidier but (nitpicking) the one above performs an implicit assertion

File: devito/passes/clusters/cse.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/passes/clusters/cse.py:116]: wdym ? we keep on looping and `processed` needs to be updated? what change are you proposing ?

Comment (EdCaunt) [reply devito/passes/clusters/cse.py:116]: Ignore me then

Comment (EdCaunt) [reply devito/passes/clusters/cse.py:116]: Missed an indent

**Review by EdCaunt — APPROVED** [review]
Comments responded, looks good

**Review by mloubout — APPROVED** [review]

**Review by mloubout — APPROVED** [review]

**Pull request merged**